fix(ui): command line issues with 'cmdheight' set to zero

luukvbaal opened the pull request:
**fix(ui): allow resize commands to set 'cmdheight' to 0**

Resolve https://github.com/neovim/neovim/issues/21558

**fix: issues with command line if ui elements are externalized**

Resolve https://github.com/neovim/neovim/issues/20888 and handle side effects for setting `'cmdheight'` to zero.

luukvbaal committed:
fix(ui): allow resize commands to set 'cmdheight' to 0

Resolve https://github.com/neovim/neovim/issues/21558

max397574 commented:
would this perhaps also fix this issue https://github.com/neovim/neovim/issues/20888 ?

justinmk commented:
does this need a test? (or just add a small tweak to an existing test)

luukvbaal commented:
> would this perhaps also fix this issue https://github.com/neovim/neovim/issues/20888 ?

I don't think so.

> does this need a test? (or just add an extra case to an existing test)

Sure, but I think we first need to decide if it's actually supposed to work as proposed. I think it makes sense to allow resizing the cmdline height back to 0 if it was set explicitly but could be debatable.

I'll look into the issue and testcase this weekend.

luukvbaal committed:
fix: issues with command line if ui elements are externalized

Resolve https://github.com/neovim/neovim/issues/20888 and handle side
effects for setting 'cmdheight' to zero.

luukvbaal committed:
test: add test cases for command line issues

zeertzjq commented on test/functional/ui/cmdline_spec.lua:
```diff
@@ -1364,4 +1363,40 @@ describe('cmdheight=0', function()
+
+  it("clears cmdline_area when resized with external messages", function()
+    clear()
+    screen = new_screen({rgb=true, ext_messages=true, ext_linegrid=linegrid})
```
```suggestion
    screen = new_screen({rgb=true, ext_messages=true})
```

luukvbaal commented:
> I'll look into the issue and testcase this weekend.

Did this the same day btw, if it wasn't clear. As far as I'm concerned, latest commit fixes both issues. Unless if someone disagrees with the proposed `'cmdheight'` resizing behavior. I see three possible behaviors:

* Never able to set `'cmdheight'` to zero through resizing.
* Always able to set `'cmdheight'` to zero through resizing.
* Only able to set `'cmdheight'` to zero through resizing if it was explicitly set to zero before.

First commit in this PR implements the third option.
